zebra-consensus/src/chain.rs

```diff
@@ -22,68 +22,87 @@ use zebra_state as zs;
 
 use crate::{
     block::BlockVerifier,
-    block::VerifyBlockError,
-    checkpoint::{CheckpointList, CheckpointVerifier, VerifyCheckpointError},
+    checkpoint::{CheckpointList, CheckpointVerifier},
     BoxError, Config,
 };
 
+/// The bound for each verifier's buffer.
+///
+/// We choose the verifier buffer bound based on the maximum number of
+/// concurrent verifier users, to avoid contention:
+///   - the `ChainSync` component
+///   - the `Inbound` service
+///   - a miner component, which we might add in future, and
+///   - 1 extra slot to avoid contention.
+///
+/// We deliberately add extra slots, because they only cost a small amount of
+/// memory, but missing slots can significantly slow down Zebra.
+const VERIFIER_BUFFER_BOUND: usize = 4;
+
 /// The chain verifier routes requests to either the checkpoint verifier or the
 /// block verifier, depending on the maximum checkpoint height.
 struct ChainVerifier<S>
 where
     S: Service<zs::Request, Response = zs::Response, Error = BoxError> + Send + Clone + 'static,
     S::Future: Send + 'static,
 {
-    block: BlockVerifier<S>,
-    checkpoint: CheckpointVerifier<S>,
+    // Normally, we erase the types on buffer-wrapped services.
+    // But if we did that here, the block and checkpoint services would be
+    // type-indistinguishable, risking future substitution errors.
+    block_verifier: Buffer<BlockVerifier<S>, Arc<block::Block>>,
+    checkpoint_verifier: Buffer<CheckpointVerifier<S>, Arc<block::Block>>,
     max_checkpoint_height: block::Height,
 }
 
 #[derive(Debug, Display, Error)]
 pub enum VerifyChainError {
     /// block could not be checkpointed
-    Checkpoint(#[source] VerifyCheckpointError),
+    Checkpoint(#[source] BoxError),
     /// block could not be verified
-    Block(#[source] VerifyBlockError),
+    Block(#[source] BoxError),
```

Comment:
Since we're using a `Buffer`, we get `BoxError`s back from the checkpoint and block verifiers.